crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1129,14 +1203,34 @@ impl<'db> TypeInferenceBuilder<'db> {
             is_async: _,
         } = comprehension;
 
+        if !is_first {
+            self.infer_expression(iter);
+        }
         self.infer_expression(target);
-        self.infer_expression(iter);
         for if_clause in ifs {
             self.infer_expression(if_clause);
         }
+    }
 
-        // TODO comprehension type
-        Type::Unknown
+    fn infer_comprehension_definition(
+        &mut self,
+        comprehension: &ast::Comprehension,
+        is_first: bool,
+        definition: Definition<'db>,
+    ) {
+        let ast::Comprehension {
+            range: _,
+            target,
+            iter,
+            ifs: _,
+            is_async: _,
+        } = comprehension;
+
+        if !is_first {
+            self.infer_expression(iter);
+        }
+        let target_ty = self.infer_expression(target);
```

Comment:
I might add a TODO here just to remind us that actually the `target` type should (later, not in this PR) be inferred by getting the contained type of the iterable type we get for `iter`, not by an `infer_expression` call.